src/App.jsx

```diff
@@ -0,0 +1,53 @@
+import React, { useState } from 'react';
+import * as R from 'ramda';
+
+import Todo from './Todo';
+
+export default function App() {
+  const [state, setState] = useState({
+    userInput: '',
+    tasks: [],
+  });
+
+  const { userInput, tasks } = state;
+
+  function handleChangeInput(newInput) {
+    setState({
+      userInput: newInput,
+      tasks,
+    });
+  }
+
+  function handleDeleteTask(target) {
+    const isTaskToRemain = R.pipe(R.prop('id'), R.equals(target), R.not);
+    setState({
+      userInput,
+      tasks: (R.filter(isTaskToRemain, tasks)),
+    });
+  }
+
+  function lastIdOf(array) {
+    return R.last(R.map(R.prop('id'), array));
+  }
+
+  function getNewId(array) {
+    return R.isEmpty(array) ? 0 : lastIdOf(array) + 1;
+  }
+
+  function handleAddTask() {
+    setState({
+      userInput: '',
+      tasks: [...tasks, { id: getNewId(tasks), content: userInput }],
+    });
+  }
+
+  return (
+    <Todo
+      userInput={userInput}
+      tasks={tasks}
+      handleChangeInput={(e) => handleChangeInput(e.target.value)}
+      handleAddTask={handleAddTask}
+      handleDeleteTask={handleDeleteTask}
```

Comment:
```suggestion
      onChangeInput={handleChangeInput}
      onAddTask={handleAddTask}
      onDeleteTask={handleDeleteTask}
```
실제로 `event.target.value` 를 넘기는 건 해당 컴포넌트에 위임하는건 어떨까요? App 컴포넌트 관점에선 `handleChangeInput` 함수는 단순히 값을 받아서 상태값을 업데이트해줄 뿐인거죠.

추가로 받는 쪽에서는 on으로 받도록 처리해주세요.

Reply:
> 실제로 event.target.value 를 넘기는 건 해당 컴포넌트에 위임하는건 어떨까요?
좋은 말씀입니다.  `6e546af`으로 수정했습니다.

Reply:
> 추가로 받는 쪽에서는 on으로 받도록 처리해주세요.

이게 조금 헷갈리는 부분인데, 저도 
```js
handleChangeInput={(e) => handleChangeInput(e.target.value)}
handleAddTask={handleAddTask}
handleDeleteTask={handleDeleteTask}
```
이렇게 둘다 `handle`로 되는게 어색했습니다.

가장 바람직한 형태는 아래와 같겠지요.

```js
// App.jsx
  return (
    <Todo
      userInput={task}
      tasks={tasks}
      onChangeInput={handleChangeInput}
      onAddTask={handleAddTask}
      onDeleteTask={handleDeleteTask}
    />
  );

// Todo.jsx
  return (
    <p>
      <h1>To-do</h1>
      <Input
        value={userInput}
        onChangeInput={handleChangeInput}
        onSubmit={handleAddTask}
      />
      <List
        items={tasks}
        onClick={handleDeleteTask}
      />
    </p>
  );
```

위와 같이 작성하는 것이 적절한 이유는 아래와 같이 이해하고 있습니다.


`App`은 상태만 관리합니다. `App`은 `task`, `tasks`, 그리고 이들을 수정하는 함수들 밖에 모릅니다. 이것들을 가지고 `Todo`를 그리는건 `Todo컴포넌트`의 일입니다.

`Todo컴포넌트`는 실제 `task`를 가지고 일을 하는 곳이니, 받을 때 `onChangeInput`, `onAddTask`, `onDeleteTask`와 같이 `구체적`으로 받습니다.

`이와 다르게` `Input`, `List 컴포넌트`는 `Todo`가 뭐하는 것인지는 모르고 받은 함수를 이벤트의 발생에 따라 실행시켜줄 뿐이므로 `onChangeInput`, `onSubmit`, `onClick`와 같이 받습니다.


그런데 이렇게 작성하려면 중간에 `Todo.jsx`에서 다음과 같이 받아야 합니다. 이름을 바꿔줘야하니까요.
```js
export default function Todo({
  userInput,
  tasks,
  onChangeInput: handleChangeInput,
  onAddTask: handleAddTask,
  onDeleteTask: handleDeleteTask,
}) 
```
이렇게 하는 것이 맞나요? 
일단은 이렇게 하는 방법밖에 없는 것 같아 `77a7c96`로 수정했습니다.

Reply:
제가 설명이 조금 부족했던 것 같아요. 자식 컴포넌트에선 받은 함수를 그대로 사용하거나 전달하기 때문에 `handle` prefix를 붙이지 않고 `on`을 그대로 사용하는게 맞을 것 같아요~

```javascript
export default function Todo({
  userInput,
  tasks,
  onChangeInput,
  onAddTask,
  onDeleteTask
}) 

```

Reply:
음 그런데 `on~`, `handle~`로 이름을 짓는 이유가 그 컴포넌트를 사용하는 것만 보고 뭐하는지 알려고 하는 것 아닌가요?
`App.jsx`에서 `Todo`를 사용할 때는 `on~`, `handle~`로 하는데, 그 하위에서는 그것이 지켜지지 않는 것은 조금 부자연스럽다고 느껴지네요.

Reply:
맞아요. 네이밍만으로 뭘 하는지 알 수 있어야합니다. on/handle은 실제로 구현된 함수인지 아니면 props로 받아온 이벤트인지를 구분합니다.

App 컴포넌트에서는 우리가 사용할 실제 함수를 `handle prefix`를 붙여서 선언한거고, 자식 컴포넌트로 props로 넘겨줬습니다.
자식 컴포넌트에서는 `on prefix`가 붙은 함수 타입 props를 받아서 해당 이벤트를 실행하거나 또다른 하위 컴포넌트로 넘기게 됩니다.

제 설명이 조금 부족했던 것 같아 잘 번역된 블로그의 글을 첨부해볼게요 :) 참고해보시면 좋을 것 같습니다 🙏🏻
의논이 더 필요한 부분이 있다면 말씀주세요 😊

> `on` 접두사가 붙은 경우, 이 `Prop`에 실제 이벤트가 연결되어 있다는 걸 뜻하고,
`handle` 접두사가 붙은 경우, 이벤트가 발생했을 때 호출되는 `실제 Function`을 의미합니다.

참조:
- https://blog.sonim1.com/220

Reply:
이해가 되었습니다. `94f7116`로 수정했습니다. 감사합니다.